File: lib/lambda/rules/aws_cis/aws_cis_3_5.ts
import { cloudTrailRecord, uguisuRule, detection } from "../../models";

// https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cis-controls.html#cis-3.5-remediation

export class rule extends uguisuRule {
  constructor() {
    super({
      id: "aws_cis_3.5",
      title: "CloudTrail configuration changes",
      description:
        "AWS CIS benchmark 3.5 recommend to ensure a log metric filter and alarm exist for CloudTrail configuration changes",
      severity: "medium",
    });
  }

  detect(record: cloudTrailRecord): detection | null {
    const targetEvents: { [key: string]: boolean } = {
      CreateTrail: true,
      UpdateTrail: true,
      DeleteTrail: true,
      StartLogging: true,
      StopLogging: true,
    };

    if (targetEvents[record.eventName] === true) {
      return {
        rule: this,
        event: record,
      };
    }

    return null;
  }
}

File: pkg/rules/aws_cis_3_5.go
package rules

import (
	"github.com/m-mizutani/uguisu/pkg/models"
)

// https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cis-controls.html#cis-3.5-remediation
type awsCIS3_5 struct {
	targetEvents map[string]bool
}

func newAwsCIS3_5() models.Rule {
	return &awsCIS3_5{
		targetEvents: map[string]bool{
			"CreateTrail":  true,
			"UpdateTrail":  true,
			"DeleteTrail":  true,
			"StartLogging": true,
			"StopLogging":  true,
		},
	}
}

func (x *awsCIS3_5) ID() string                { return "aws_cis_3.5" }
func (x *awsCIS3_5) Title() string             { return "CloudTrail configuration changes" }
func (x *awsCIS3_5) Severity() models.Severity { return models.SeverityMedium }
func (x *awsCIS3_5) Description() string {
	return "AWS CIS benchmark 3.5 recommend to ensure a log metric filter and alarm exist for CloudTrail configuration changes"
}
func (x *awsCIS3_5) Match(record *models.CloudTrailRecord) bool {
	return x.targetEvents[record.EventName]
}
